### PR TITLE
session/replaytest: add multi-backend replay consistency test framework

### PR DESCRIPTION
**Pull request by Stelquis**
Fixes #2001

新增 `session/replaytest/` 包，提供多后端回放一致性测试框架。用同一组标准化操作序列驱动多个后端，自动
逐字段比较并生成差异报告，精确定位到 session id、event index、summary filter-key、track name、memory id
及字段路径。

### 核心组件

- **ReplayOp / ReplayCase** — 14 种操作类型，声明式定义回放序列
- **Harness** — 编排多后端执行 → 归一化 → 比较 → 输出报告
- **Normalizer** — ID 占位符、UTC ±1s、JSON 排序、map 排序、浮点 ±0.01 容差、丢弃私有 metadata
- **Comparator** — 逐字段比较 Event/State/Memory/Summary/Track，基于后端能力矩阵标记 AllowedDiff
- **Reporter** — JSON + 文本双格式差异报告
- **TrapInjector** — 7 种预置异常注入策略，验证框架检出能力
- **MockModel** — 固定 seed 生成可重复的多轮对话 + tool call 事件序列

### 10 条 Replay Case

覆盖单轮/多轮对话、工具调用、State 更新、Memory 读写、Summary 生成与截断（含 filter-key）、Track 事件
、并发写入、异常恢复。

### 后端接入

后端通过 `BackendFactory` 注册，环境变量控制开关。默认启用 InMemory + SQLite，可选 Redis/Postgres/MyS
QL/ClickHouse。

### 附带修复

- 修复 `TestNormalizer_NormalizeSession` 中 map 遍历顺序非确定性问题（sort.Strings）
- Memory 比较改为按内容排序后匹配（集合相等语义），而非按返回顺序逐位置比对
- `TrapShiftTimestamp` 移除 `-1` 哨兵值
- Case4 改用 `OpDeleteSessionState` 测试真实删除语义
- Case9 改用 goroutine 实现真正并发写入
- `OpGetSessionSummaryText` 支持 filter-key 参数传递
- `MockModel.GenerateConversation` 增加 nil-safety 保护

### 依赖升级
| 依赖 | 原版本 | 目标版本 | 原因 |                                                                    
|------|--------|---------|------|                                                                     
| google.golang.org/grpc | v1.65.0 | v1.79.3 | CVE 修复 |                                              
| github.com/getkin/kin-openapi | v0.124.0 | v0.131.0 | CVE 修复 |                                     
| golang.org/x/crypto | v0.32.0 | v0.52.0 | CVE 修复 |                                                 
                                                                                                     
### 验证                                                                                               
                                                                                                     
```bash                                                                                                
go build ./session/replaytest/...    # 通过                                                            
go test ./session/replaytest/...     # 通过（0.08s，115+ 测试）                                        
                                   
全 Mock，无需 API Key。零改动现有模块。                                                                  
                        
RELEASE NOTES: 新增 session/replaytest 多后端回放一致性测试框架；修复 map 遍历测试 flaky；新增并发写入
测试；升级漏洞依赖。

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -1,6 +1,6 @@
 module trpc.group/trpc-go/trpc-agent-go
 
-go 1.21
+go 1.25.0
```
Comment (Flash-LHR) [inline go.mod]: Raising the root module to Go 1.25 makes `go test ./...` fail under the new vet rules. Keep `go 1.21` or fix the vet failure before this upgrade.
<details><summary>中文</summary>
将根模块升到 Go 1.25 会触发新 vet 规则并导致 `go test ./...` 失败。请保留 `go 1.21`，或先修复 vet 失败再升级。
</details>

File: session/replaytest/harness.go [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+			}
+			res.BackendName = b.Name
+			res.Duration = elapsed
+			results[b.Name] = res
+		}
+
+		allResults = append(allResults, caseResult{
+			caseName: c.Name,
+			results:  results,
+			errors:   errs,
```
Comment (Flash-LHR) [inline session/replaytest/harness.go:105]: Stored backend errors are never surfaced, so `Run` can return a clean report when a backend fails a case. Return those errors or report them as diffs.
<details><summary>中文</summary>
保存的后端错误从未对外暴露，因此某个后端执行用例失败时 `Run` 会返回干净报告。请返回这些错误，或把它们写入差异报告。
</details>

File: session/replaytest/backends.go [hunk 3]
```diff
@@ -0,0 +1,625 @@
...
+}
+
+// init registers the built-in backends.
+func init() {
+	// InMemory backend — always enabled by default.
+	RegisterBackend(BackendFactory{
+		Name:    "InMemory",
+		Enabled: envEnabled("INMEMORY", true),
+		New: func() (session.Service, memory.Service, error) {
+			sessSvc := inmemorysession.NewSessionService()
```
Comment (Rememorio) [inline session/replaytest/backends.go:220]: **P1: Configure a deterministic summarizer for replay backends**

The built-in session services are created without `WithSummarizer`. Existing session implementations return immediately from `CreateSessionSummary` when no summarizer is configured, so cases 6 and 7 compare empty summary maps and `GetSessionSummaryText` returns no value; they cannot detect summary persistence or filter-key regressions.

Inject the same deterministic `SessionSummarizer` into every replay backend, or make the factory accept one. The fixtures should assert nonempty summaries for both the default and branch filter keys before comparing backends.

<details><summary>中文</summary>

**为回放后端配置确定性 Summarizer**

内置 session 服务创建时都没有传入 `WithSummarizer`。现有实现未配置 summarizer 时，`CreateSessionSummary` 会直接返回，因此 case 6 和 case 7 实际比较的是空 Summary map，`GetSessionSummaryText` 也取不到内容，无法发现 Summary 持久化或 filter-key 行为回归。

建议为所有回放后端注入同一个确定性 `SessionSummarizer`，或让工厂接收该依赖。跨后端比较前，fixture 应先断言默认 key 和分支 key 都生成了非空 Summary。

</details>

File: session/replaytest/backends.go [hunk 4]
```diff
@@ -0,0 +1,625 @@
...
+		}
+		if err := sessSvc.UpdateSessionState(ctx, op.Key, sd.State); err != nil {
+			return fmt.Errorf("UpdateSessionState: %w", err)
+		}
+
+	case OpDeleteSessionState:
+		// Only the first key is deleted for simplicity.
+		if stateMap, ok := op.Data.(session.StateMap); ok {
+			for key := range stateMap {
+				if err := sessSvc.UpdateSessionState(ctx, op.Key, session.StateMap{key: nil}); err != nil {
```
Comment (Rememorio) [inline session/replaytest/backends.go:400]: **P1: Do not label a nil assignment as state deletion**

`UpdateSessionState(..., {key: nil})` does not delete the key under the existing service contract. Current implementations persist a present key with a nil value, and `Session.GetState` still reports that key as present, so case 4 can pass across backends without testing deletion semantics.

There is no `DeleteSessionState` method on `session.Service`. Either remove or rename this replay operation to describe nil assignment, or drive a genuine deletion path and assert that the key is absent after `GetSession`.

<details><summary>中文</summary>

**不要把 nil 赋值当作 State 删除**

按照现有服务契约，`UpdateSessionState(..., {key: nil})` 不会删除该 key。当前实现会持久化一个值为 nil、但仍然存在的 key，`Session.GetState` 也会报告它存在，因此 case 4 可以在所有后端一致通过，却完全没有验证删除语义。

`session.Service` 当前并不存在 `DeleteSessionState` 方法。应删除或重命名这个回放操作以准确表达 nil 赋值，或者改用真正的删除路径，并在 `GetSession` 后断言该 key 已不存在。

</details>

File: session/replaytest/backends.go [hunk 5]
```diff
@@ -0,0 +1,625 @@
...
+	case OpUpdateMemory:
+		md, ok := op.Data.(MemoryData)
+		if !ok {
+			return fmt.Errorf("UpdateMemory: invalid data type %T", op.Data)
+		}
+		var opts []memory.UpdateOption
+		if md.Metadata != nil {
+			opts = append(opts, memory.WithUpdateMetadata(md.Metadata))
+		}
+		memKey := memory.Key{AppName: md.UserKey.AppName, UserID: md.UserKey.UserID, MemoryID: ""}
```
Comment (Rememorio) [inline session/replaytest/backends.go:429]: **P1: Provide a usable memory ID for updates**

Every valid `OpUpdateMemory` constructs a `memory.Key` with an empty `MemoryID`. All current memory implementations validate this key and return `ErrMemoryIDRequired`, while `MemoryData` provides no way for a caller to supply or resolve the ID, so this public operation cannot succeed.

Define an update payload with a memory ID or symbolic handle and capture each backend's effective ID from `AddMemory`. Add an add-update-read case that proves the update succeeds on at least InMemory and SQLite.

<details><summary>中文</summary>

**为 Memory 更新提供可用的 ID**

每次执行有效的 `OpUpdateMemory` 都会构造一个 `MemoryID` 为空的 `memory.Key`。现有 memory 实现都会校验该字段并返回 `ErrMemoryIDRequired`，而 `MemoryData` 又没有提供传入或解析 ID 的方式，因此这个公开操作实际上无法成功。

建议为更新操作定义包含 Memory ID 或符号句柄的 payload，并在 `AddMemory` 后记录各后端实际生成的 ID。请增加 add-update-read 场景，至少验证 InMemory 和 SQLite 能成功更新。

</details>

File: session/replaytest/backends.go [hunk 6]
```diff
@@ -0,0 +1,625 @@
...
+			return fmt.Errorf("ReadMemories: invalid data type %T", op.Data)
+		}
+		entries, err := memSvc.ReadMemories(ctx, uk, 100)
+		if err != nil {
+			return fmt.Errorf("ReadMemories: %w", err)
+		}
+		result.Memories = entries
+
+	case OpSearchMemories:
+		type searchData struct {
```
Comment (Rememorio) [inline session/replaytest/backends.go:542]: **P1: Expose the search query in the replay contract**

`searchData` is a function-local defined type, so callers cannot construct a value that satisfies this type assertion. The only reachable fallback accepts `memory.UserKey` and always searches with an empty query, meaning `OpSearchMemories` and case 5 do not exercise query matching or ranking semantics.

Expose a payload containing `UserKey` and `Query`, or otherwise add the query to the public replay operation contract. Use a nonempty fixture query that excludes at least one stored memory.

<details><summary>中文</summary>

**在回放契约中公开搜索 Query**

`searchData` 是函数内部定义的具名类型，调用方无法构造能通过该类型断言的值。唯一可达的 fallback 只接收 `memory.UserKey`，并始终使用空 Query，因此 `OpSearchMemories` 和 case 5 都没有验证搜索匹配或排序语义。

建议公开一个包含 `UserKey` 与 `Query` 的 payload，或以其他方式把 Query 纳入公开的回放操作契约。fixture 应使用非空 Query，并确保至少排除一条已存储的 Memory。

</details>

File: session/replaytest/comparator.go [hunk 7]
```diff
@@ -0,0 +1,816 @@
...
+	diffs = append(diffs, c.compareString(caseName, a, b, "session.userID", sessA.UserID, sessB.UserID)...)
+
+	// Compare events.
+	diffs = append(diffs, c.compareEventLists(caseName, a, b, sessA.Events, sessB.Events)...)
+
+	// Compare state.
+	diffs = append(diffs, c.compareStateMaps(caseName, a, b, "session.state", sessA.State, sessB.State)...)
+
+	// Compare summaries.
+	diffs = append(diffs, c.compareSummaryMaps(caseName, a, b, "session.summaries", sessA.Summaries, sessB.Summaries)...)
```
Comment (Rememorio) [inline session/replaytest/comparator.go:167]: **P1: Compare tracks from the session snapshot**

`NormalizeSession` retains `Session.Tracks`, but `compareSessions` never compares them. The separate `compareTracks` path reads `BackendResult.Tracks`, which `executeOps` initializes as empty and never populates, so case 8 can pass even if one backend loses every persisted track event.

Compare `sessA.Tracks` and `sessB.Tracks` here, or populate `BackendResult.Tracks` from the refreshed session before comparison. Add a regression test where the two session snapshots differ in track count or payload.

<details><summary>中文</summary>

**比较 Session 快照中的 Track**

`NormalizeSession` 会保留 `Session.Tracks`，但 `compareSessions` 没有比较它。另一条 `compareTracks` 路径读取的是 `BackendResult.Tracks`，而 `executeOps` 只初始化了这个 map，之后从未填充，因此即使某个后端丢失全部 Track 事件，case 8 仍可能通过。

可以直接在这里比较 `sessA.Tracks` 与 `sessB.Tracks`，或在刷新 Session 后把 Track 写入 `BackendResult`。请增加两个 Session 快照在 Track 数量或 Payload 上不一致的回归测试。

</details>

File: session/replaytest/comparator.go [hunk 8]
```diff
@@ -0,0 +1,816 @@
...
+
+	return diffs
+}
+
+// compareEvent compares two individual events field by field.
+func (c *Comparator) compareEvent(caseName string, a, b *BackendResult, prefix string, evA, evB event.Event) []DiffEntry {
+	var diffs []DiffEntry
+
+	// Get the first choice message for content/role comparison.
+	msgA := firstMessage(&evA)
```
Comment (Rememorio) [inline session/replaytest/comparator.go:207]: **P1: Compare the full persisted event contract**

Using `firstMessage` limits comparison to `Choices[0].Message`. Additional choices and persisted response fields such as `Error`, `Done`, `Model`, and `Usage` are ignored; event correlation fields are also omitted, and the normalizer collapses every nonempty invocation ID to the same placeholder. A backend that drops or corrupts these fields can still be reported as consistent.

Define a canonical comparison for the full persisted `event.Event`: preserve semantic correlation relationships, compare all choices and relevant response/event fields, and normalize only explicitly generated or backend-private values. Add traps for a second choice, response error/done state, and parent invocation metadata.

<details><summary>中文</summary>

**比较完整的持久化 Event 契约**

通过 `firstMessage` 比较时，只会检查 `Choices[0].Message`。其他 Choice，以及 `Error`、`Done`、`Model`、`Usage` 等持久化 Response 字段都被忽略；事件关联字段也没有参与比较，Normalizer 还会把所有非空 Invocation ID 压成同一个占位符。因此后端即使丢失或破坏这些字段，仍可能被报告为一致。

建议为完整的持久化 `event.Event` 定义规范化比较：保留语义关联关系，比较所有 Choice 以及相关 Response/Event 字段，只归一化明确属于自动生成或后端私有的值。请增加第二个 Choice、Response error/done 状态和 Parent Invocation metadata 的异常注入测试。

</details>

File: session/replaytest/fixtures.go [hunk 9]
```diff
@@ -0,0 +1,355 @@
...
+		Case8_TrackEvents(),
+		Case9_ConcurrentWrites(),
+		Case10_Idempotency(),
+	}
+}
+
+// Case1_SingleTurn covers basic session creation and event appending.
+// Trap verification point: event content/order/role tampering.
+func Case1_SingleTurn() ReplayCase {
+	key := session.Key{AppName: "fixture", UserID: "user1", SessionID: "case1"}
```
Comment (Rememorio) [inline session/replaytest/fixtures.go:40]: **P1: Isolate persistent backend data per run**

The fixtures reuse fixed app, user, and session IDs, while the harness only closes services and never removes their data. A second run against Postgres or MySQL can fail with an existing-session error; Redis can reuse the old session and append another copy of the events. Parallel runs against the same configured database also collide.

Generate a unique run namespace while keeping keys identical across backends within that run, or delete all created sessions and memories during teardown. Add a repeated-run scenario against a persistent backend or persistence-capable fake.

<details><summary>中文</summary>

**隔离每次运行的持久化后端数据**

这些 fixture 固定复用 App、User 和 Session ID，而 Harness 结束时只关闭服务，从不清理数据。对 Postgres 或 MySQL 重复执行时可能直接因 Session 已存在而失败；Redis 则可能复用旧 Session 并再次追加事件。多个任务共用同一数据库并行执行时也会相互冲突。

建议为每次运行生成唯一命名空间，同时保证同一次运行内各后端使用相同 key；或者在 teardown 中删除所有创建的 Session 和 Memory。请增加一个针对持久化后端或持久化 fake 的重复执行场景。

</details>

File: session/replaytest/harness.go [hunk 10]
```diff
@@ -0,0 +1,286 @@
...
+				errs[b.Name] = fmt.Errorf("create services: %w", err)
+				continue
+			}
+
+			start := time.Now()
+			res, err := executeOps(ctx, sessSvc, memSvc, c.Ops)
+			elapsed := time.Since(start)
+
+			sessSvc.Close()
+			memSvc.Close()
```
Comment (Rememorio) [inline session/replaytest/harness.go:91]: **P1: Handle the documented nil memory service**

The ClickHouse factory deliberately returns a nil `memory.Service`, but this line closes it unconditionally; `RunBackends` and `RunTrap` have the same assumption. Enabling the documented ClickHouse backend therefore panics after the first session-only case, and memory operations can dereference it before capability handling occurs.

Model unsupported services explicitly: guard lifecycle calls and skip or report unsupported operations based on capabilities, or return a safe no-op service. Add a ClickHouse-like factory test that returns a session service with nil memory service.

<details><summary>中文</summary>

**处理文档允许的 nil memory 服务**

ClickHouse 工厂明确返回 nil `memory.Service`，但这里无条件调用 `Close`；`RunBackends` 和 `RunTrap` 也有相同假设。因此启用文档中的 ClickHouse 后端后，即使只执行 session 场景，也会在收尾阶段 panic；memory 操作则可能更早解引用 nil。

应显式建模“不支持某项服务”：生命周期调用需要判空，并依据能力跳过或报告不支持的操作；也可以提供安全的 no-op 服务。请增加一个只返回 session 服务、memory 服务为 nil 的工厂测试。

</details>

File: session/replaytest/harness.go [hunk 11]
```diff
@@ -0,0 +1,286 @@
...
+
+			if err != nil {
+				errs[b.Name] = err
```
Comment (Rememorio) [inline session/replaytest/harness.go:94]: **P1: Do not discard backend execution failures**

Execution failures are recorded in `errs`, but `caseResult.errors` is never read. If a backend cannot be created or an operation fails, that backend is simply absent from `cr.results`; with only one successful backend, the case is omitted entirely, and `Run` still returns a report with a nil error and zero unallowed diffs.

Please surface every backend/case failure as either an aggregated return error or an explicit unallowed report entry. A test with one failing factory or operation should prove that `Run` cannot report consistency after a backend failure.

<details><summary>中文</summary>

**不要丢弃后端执行错误**

执行错误虽然写入了 `errs`，但 `caseResult.errors` 后续从未被读取。某个后端创建失败或执行操作失败时，它只会从 `cr.results` 中消失；如果仅剩一个成功后端，该 case 甚至不会进入报告，而 `Run` 仍可能返回 nil error 和零个非允许差异。

建议将每个 backend/case 的失败汇总为返回错误，或生成明确的非允许差异。请增加一个后端工厂或操作失败的测试，确保这种情况下绝不会报告“后端一致”。

</details>

File: session/replaytest/backends.go [hunk 12]
```diff
@@ -0,0 +1,625 @@
...
+		}
+		cd, ok := op.Data.(ConcurrentEventData)
+		if !ok {
+			return fmt.Errorf("ConcurrentAppendEvents: invalid data type %T", op.Data)
+		}
+		var wg sync.WaitGroup
+		errCh := make(chan error, len(cd.Events))
+		for _, e := range cd.Events {
+			wg.Add(1)
+			go func(ev *event.Event) {
```
Comment (Rememorio) [inline session/replaytest/backends.go:515]: **P2: Do not compare scheduler-dependent event order**

Each backend execution launches a separate goroutine race, so there is no backend-independent ordering for these events. The comparator is position-sensitive, and the added tests explicitly tolerate or only log case 9 differences; this scenario therefore measures scheduler timing and can hide a real missing or duplicated event among accepted differences.

For this operation, compare an event multiset and assert no loss or duplication, or attach a shared deterministic ordering token and sort by it before comparison. The test should accept only order variance while still failing on content or count changes.

<details><summary>中文</summary>

**不要比较由调度器决定的事件顺序**

每个后端都会独立启动一组存在竞争关系的 goroutine，因此这些事件没有跨后端统一的顺序。Comparator 又按位置比较，而新增测试对 case 9 的差异选择容忍或仅记录日志；结果是该场景主要测到了调度时机，并可能把真实的事件丢失或重复混在被接受的差异中。

对于该操作，可以按事件集合比较并严格断言没有丢失或重复；或者为所有后端附加同一个确定性排序标记，再按该标记排序。测试应只容忍顺序变化，内容或数量变化仍必须失败。

</details>

**Review by Rememorio — COMMENTED** [review]
I reviewed the changed lines and left 10 inline comments below. The comments focus on issues worth addressing before merge.

Fresh review found several high-confidence false-pass and unusable-operation paths. Backend execution failures are discarded, ClickHouse can panic on its documented nil memory service, and the summary, track, memory-update, memory-search, and state-deletion scenarios do not exercise their stated contracts. The comparator also omits important persisted event fields, while fixed fixture keys and nondeterministic concurrent ordering make external-backend runs unreliable. Validation should include failing/partial factories, nonempty summaries and tracks, successful update/search/delete operations, repeated persistent-backend runs, and corruption traps for the full event contract.

<details><summary>中文</summary>

我看过这次变更的相关代码，在下面留下 10 条行内评论。评论聚焦在合并前值得处理的问题。

本次审查发现了多处可能导致误判通过或操作不可用的高置信度问题：后端执行错误被丢弃，ClickHouse 按文档返回 nil memory 服务时会触发 panic；Summary、Track、Memory 更新与搜索、State 删除等场景也没有真正验证其声明的契约。Comparator 还遗漏了重要的持久化事件字段，固定的 fixture key 与无确定顺序的并发写入则会降低外部后端测试的可靠性。建议重点补充失败后端、非空 Summary/Track、成功的更新/搜索/删除、持久化后端重复执行，以及完整事件字段异常注入等验证。

</details>

**Comment by liuzengh** [comment]
感谢你对本 Issue 的贡献！

犀牛鸟活动的 Issue 提交阶段已经结束，本 PR 将关闭。再次感谢你的参与与付出！
